File: python/2020/day_13.py
from math import lcm
from pathlib import Path


def main():
    text = (Path(__file__).parent / "../../input/2020/input_13.txt").read_text()
    # Part 1
    start, timeline = text.splitlines()
    start = int(start)
    bus_data = {int(b): i for i, b in enumerate(timeline.split(',')) if b != 'x'}
    time_map = {b: start - start % b + b for b in bus_data}
    first = min(time_map, key=time_map.get)
    print("Part 1:", first * (time_map[first] - start))
    # Part 2
    time = 0
    period = 1
    for bus, offset in bus_data.items():
        while (time + offset) % bus != 0:
            time += period
        period = lcm(period, bus)
    print("Part 2:", time)


if __name__ == "__main__":
    main()

File: rust/2020/src/bin/day_13.rs
use std::collections::HashMap;
use std::fs;

fn main() {
    let input = fs::read_to_string("../input/input_13.txt").unwrap();
    let lines: Vec<_> = input.lines().collect();
    // Part 1
    let start: u64 = lines[0].parse().unwrap();
    let bus_data: HashMap<u64, u64> = lines[1].split(',')
        .enumerate()
        .filter_map(|(i, b)| match b.parse() {
            Ok(x) => Some((x, i as u64)),
            Err(_) => None,
        })
        .collect();
    let first = bus_data.keys().min_by_key(|&&bus| start - start % bus + bus).unwrap();
    println!("Part 1: {}", first * (first - start % first));
    // Part 2
    let (mut time, mut period) = (0, 1);
    for (bus, offset) in bus_data {
        while (time + offset) % bus != 0 {
            time += period;
        }
        period *= bus;
    }
    println!("Part 2: {}", time);
}
